src/sentry/tasks/merge.py

```diff
@@ -301,18 +301,31 @@ def merge_objects(models, group, new_group, limit=1000, logger=None, transaction
     has_more = False
     for model in models:
         all_fields = model._meta.get_all_field_names()
+
+        # not all models have a 'project' or 'project_id' field, but we make a best effort
+        # to filter on one if it is available
+        has_project = 'project' in all_fields
```

Comment:
In what case will there be `project` without `project_id`. If there’s `project_id` that should always be prioritized. And regardless, since we’re worried about shards, we should explicitly only handle the column `project_id`. 

Alternatively, is it better to maybe just check if the models are from tagstore?

Reply:
See `has_group` right below in the existing code. It's checking *Django* field names.

Some models have `project` `ForeignKey` and some have a `project_id` `BigIntegerField` --- either way they become `project_id` in Postgres, but that's now how we easily "detect" and use them (by field name) in the ORM.

Reply:
To make it more explicit:

```
In [4]: [f for f in Activity._meta.get_all_field_names() if f.startswith('proj')]
Out[4]: ['project']
```

```
In [6]: [f for f in EventMapping._meta.get_all_field_names() if f.startswith('proj')]
   ...: 
Out[6]: ['project_id']
```

Reply:
Ouch, I wasn't aware of that.

So if that's the case, you should still be able to use `model.project_id = e.project_id` even if there is only a `project` field name.

The reason I'm pushing on this is because doing `e.project` is typically invoking another query just to resolve the full model. And this wouldn't be great within our deletion code.

Reply:
So would this work:

```python
has_project = 'project_id' in all_fields or 'project' in all_fields
if has_project:
    project_qs = model.objects.filter(project_id=group.project_id)
else:
    project_qs = model.objects.all()
```